src/components/Table/EditColumnsButton.tsx

```diff
@@ -0,0 +1,93 @@
+import { Dispatch, SetStateAction, useCallback, useEffect, useMemo, useRef, useState } from "react";
+import { useMenuTrigger } from "react-aria";
+import { useMenuTriggerState } from "react-stately";
+import { CheckboxGroup, CheckboxGroupItemOption } from "src/inputs";
+import { Css } from "../../Css";
+import { useTestIds } from "../../utils";
+import { Button } from "../Button";
+import { isIconButton, isTextButton, OverlayTrigger, OverlayTriggerProps } from "../internal/OverlayTrigger";
+import { GridColumn, Kinded } from "./GridTable";
+import { GridColumnState } from "./useColumns";
+
+interface EditColumnsButtonProps<R extends Kinded, S>
+  extends Pick<OverlayTriggerProps, "trigger" | "placement" | "disabled" | "tooltip"> {
+  columns: GridColumn<R, S>[];
+  setColumns: Dispatch<SetStateAction<GridColumnState<R, S>>>;
+  title?: string;
+  // for storybook purposes
+  defaultOpen?: boolean;
+}
+
+export function EditColumnsButton<R extends Kinded, S = {}>(props: EditColumnsButtonProps<R, S>) {
+  const { defaultOpen, disabled, columns, setColumns, trigger, title } = props;
+  const state = useMenuTriggerState({ isOpen: defaultOpen });
+  const buttonRef = useRef<HTMLButtonElement>(null);
+  const { menuTriggerProps } = useMenuTrigger({ isDisabled: !!disabled }, state, buttonRef);
+  const tid = useTestIds(
+    props,
+    isTextButton(trigger) ? trigger.label : isIconButton(trigger) ? trigger.icon : trigger.name,
+  );
+
+  const editableColumns = useMemo(
+    () =>
+      columns.filter((column) => {
+        if (!column.canHide) return false;
+        if (!column.name || column.name.length === 0) {
+          console.warn("Column is missing 'name' property required by the Edit Columns button", column);
+          return false;
+        }
+        return true;
+      }),
+    [columns],
+  );
+
+  const selectedColumns: string[] = editableColumns
+    .map((column) => {
+      if (column.canHide && column.visible) return column.name;
+    })
+    .filter((column) => !!column) as string[];
+
+  const options: CheckboxGroupItemOption[] = editableColumns.map((column) => ({
+    label: column.name!,
+    value: column.name!,
+  }));
```

Comment:
**suggestion:** Can we memoize the options? Maybe we when do the `useMemo` above to get `editableColumns`, we can also do this map and return (1) The selected columns and (2) the options. We dont' use `editableColumns` anywhere else directly, so the current memoization logic never really helps us. We just use it to build out `selectedColumns` and `options`, which will now appear as different lists every time we re-render, causing the whole thing to rerender. Granted, its not a big deal with this small of a component, but we want to make sure we memoize in the right place. Sorry if I led you in the wrong direction. This has changed quite a bit, so some of my previous comments may no longer be applicable. Still, I do think we want to memoize `options` and `selectedColumns`. I'd suggest something like:
```
const { options, selectedColumns } = useMemo(() => {
  ...
}, [columns]);
```